tests/unittests/test_runs/test_fallback_datasource.py

```diff
@@ -0,0 +1,240 @@
+import os
+import re
+import json
+from cloudinit.sources import DataSourceConfigDrive as dscd
+from cloudinit.sources import DataSourceOpenStack as dsos
+from cloudinit.sources import DataSourceOVF as dsovf
+from cloudinit import helpers
+from cloudinit.tests import helpers as tests_helpers
+from cloudinit.tests.helpers import mock
+from cloudinit import util
+from cloudinit.tests.helpers import populate_dir
+from cloudinit.sources.helpers import openstack
+from cloudinit import stages
+from cloudinit import safeyaml
+import httpretty as hp
+from io import StringIO
+from urllib.parse import urlparse
+
+
+OSTACK_META = {
+    'availability_zone': 'nova',
+    'hostname': 'sm-foo-test.novalocal',
+    'meta': {'dsmode': 'local', 'my-meta': 'my-value'},
+    'name': 'sm-foo-test',
+    'uuid': 'b0fa911b-69d4-4476-bbe2-1c92bff6535c'}
+
+
+OS_FILES = {
+    'openstack/latest/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/latest/network_data.json': json.dumps(
+        {'links': [], 'networks': [], 'services': []}),
+}
+
+
+OVF_ENV_CONTENT = """<?xml version="1.0" encoding="UTF-8"?>
+<Environment xmlns="http://schemas.dmtf.org/ovf/environment/1"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns:oe="http://schemas.dmtf.org/ovf/environment/1"
+  xsi:schemaLocation="http://schemas.dmtf.org/ovf/environment/1 ../dsp8027.xsd"
+  oe:id="WebTier">
+  <!-- Information about hypervisor platform -->
+  <oe:PlatformSection>
+      <Kind>ESX Server</Kind>
+      <Version>3.0.1</Version>
+      <Vendor>VMware, Inc.</Vendor>
+      <Locale>en_US</Locale>
+  </oe:PlatformSection>
+  <!--- Properties defined for this virtual machine -->
+  <PropertySection>
+{properties}
+  </PropertySection>
+</Environment>
+"""
+
+CFG_DRIVE_FILES_V2 = {
+    'openstack/latest/meta_data.json': json.dumps(OSTACK_META)}
+
+BASE_URL = "http://169.254.169.254"
+
+EC2_VERSIONS = [
+    'latest',
+]
```

Comment:
These should all be imported from unittest classes you've copied them from.  No need to duplicate this.

Reply:
Addressed in latest commit